migrations/20230613116515_courses.sql

```diff
@@ -0,0 +1,103 @@
+-- Courses Table
+CREATE TABLE IF NOT EXISTS public.courses (
+    id SERIAL PRIMARY KEY,
+    catalog_number SMALLINT NOT NULL,
+    subject_code VARCHAR(10) NOT NULL,
+    external_id VARCHAR(10) NOT NULL,
+    academic_level VARCHAR(13) NOT NULL,
+    title TEXT NOT NULL,
+    description TEXT NOT NULL,
+    abbreviated_description TEXT NOT NULL,
+    requirements_description TEXT NOT NULL,
+    prerequisites_id INTEGER NULL,
+    CONSTRAINT uk_courses_external_id UNIQUE (external_id),
+    CONSTRAINT fk_courses_to_prerequisites FOREIGN KEY (prerequisites_id)
+    REFERENCES prerequisites (id),
+    CONSTRAINT chk_courses_external_id_not_empty CHECK (external_id <> ''),
+    CONSTRAINT chk_courses_subject_code_not_empty CHECK (subject_code <> ''),
+    CONSTRAINT chk_courses_academic_level CHECK (academic_level IN ('undergraduate', 'graduate')),
+    CONSTRAINT chk_courses_title_not_empty CHECK (title <> '')
+);
+
+-- Courses Table Indices
+CREATE INDEX CONCURRENTLY IF NOT EXISTS idx_courses_subject_code_catalog_number
+    ON courses (subject_code, catalog_number);
+
+CREATE INDEX CONCURRENTLY IF NOT EXISTS idx_courses_title
+    ON courses (title);
+
+CREATE INDEX CONCURRENTLY IF NOT EXISTS idx_courses_academic_level
+    ON courses (academic_level);
+
+-- Course Offerings Table
+CREATE TABLE IF NOT EXISTS course_offerings (
+    id SERIAL PRIMARY KEY,
+    course_id INTEGER NOT NULL,
+    year SMALLINT NOT NULL,
+    term VARCHAR(6) NOT NULL,
```

Comment:
I think the year might still be helpful if we want to search up courses by the term and/or year for historical course offerings?

Reply:
Yeah, term codes do that:

> Your academic record is recorded by various terms. You may think of terms as Fall, Winter, or Spring. In the UWaterloo model, terms are often designated by a 4-digit term code "ayym" where "a" is either 0 (for years before 2000) or a 1 (for years 2000 and later) "yy" is the least significant, two digits of the year and "m" is the month that the term starts. As a result 1199 starts in the ninth month of 2019 (September or Fall); another example, 1201 starts in the first month of 2020 (January or Winter).

Reply:
Right, but then we'd have to do application-level mapping to term codes. IMO it's much easier to keep the column and query by the year value. It makes it a lot easier for testing and when we have to inspect the data manually. What do you think?

Reply:
> Our data model shouldn't be fully constrained by specific encoding formats by UW

I agree with the sentiment in general, but we would be duplicating the same data in this case. Also, we'd need an application level mapping to compute the year from the term code even if we decide to have a year attribute (since the API only provides the term code).

It's quite easy to test even when we only have the term code attribute. Since the term code is always going to be a string of length 4, a search for courses during a year (for example 2022) is equal to a quite simple regex `"122."`

Reply:
I'm actually thinking: where would we need to use the term code?

If we really need to keep the term code, then it makes more sense to keep term + year columns in the table and encode/map it to the term code in the application layer only to reduce redundant data. Well normalized schemas tend to keep the base data (building blocks) in data models and project aggregations according to the base data.

For instance, `SELECT * WHERE year > 2000` is much simpler, likely to be correct, and *readable* than selecting based on a more complex regular expression in the database.

Reply:
Students only search for courses offered in a particular term, especially due to their availability. A 1A student picking electives in the Fall doesn't want to see courses for the entire year, etc. In general, course offerings are per term and not per year. I'm actually not sure where we would need to use year?

Reply:
Discussed offline: using enums instead